Disable dynamo on functional optims if capturable=False

mlazos opened the pull request:
This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now fallback to eager if capturable=False.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze

Note on the actual fallback logic - there was an issue with torchscript originally not handling *args, **kwargs properly, after rectifying that by using `functools.wraps`, there was an additional bug with scoping which required the single tensor implementation to be in the global scope at the time of the fallback closure being created. I pass in the single tensor function to the `_disable_dynamo_if_unsupported` decorator to workaround this bug. 

Note on long term torchscript support - due to the issue above, Alban and I decided it would be a good idea to deprecate torchscript support for the functional optimizers since no one is actively maintaining it. The plan here is to eventually remove the `torch.script.jit` decorator from each distrib.optim.* functional optimizer. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123469
* __->__ #123619



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

mlazos committed:
Move step to cpu if capturable/fused is not specified

[ghstack-poisoned]

mlazos committed:
Update on "Move step to cpu if capturable/fused is not specified"




[ghstack-poisoned]

mlazos committed:
Update on "Move step to cpu if capturable/fused is not specified"




[ghstack-poisoned]

janeyx99 reviewed (commented):
Could you write a unit test in TestOptimRenewed showing the case that this PR fixes? Is the invariant we are pushing for here "step must be on CPU if not capturable or fused"?

mlazos commented:
> Could you write a unit test in TestOptimRenewed showing the case that this PR fixes? Is the invariant we are pushing for here "step must be on CPU if not capturable or fused"?

Yes, this fixes test_load_older_state dict for dynamo, let me add an extra check in that test for this invariant if that's okay.

mlazos committed:
Update on "Move step to cpu if capturable/fused is not specified"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path.




[ghstack-poisoned]

janeyx99 reviewed (commented):
Ah but why do we want this invariant? What was the bug when step was on CUDA? I'd expect that it would just be slower but it shouldn't break anything.

The concern here stems from the BC-breakingness of it if the user actually wants the step to still be on CUDA.

mlazos commented:
The bug is that we could end up running the non-capturable path with CUDA tensors, if we're compiling that can break some for each ops where scalars are expected to be on cpu. Specifically at the bottom of Adamax, we stack the scalars, addcdiv doesn't accept stacked cuda tensors as the scalar arg. 

IMO this is a bug, I think it should be fixed regardless of above. Silent slowness is worse than breaking BC here I think.

However if you don't want to break BC we can just add an is_compiling in the optimizer code to break up the addcdiv on the non captureable path. Wdyt?

mlazos committed:
Update on "Move step to cpu if capturable/fused is not specified"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path.




[ghstack-poisoned]

mlazos commented:
After some discussion offline, I opted to make this a change that will only affect compiled optimizers - basically we move the step_size to cpu if it's a tensor. This only occurs if we're compiling.

mlazos committed:
Update on "Move step to cpu if capturable/fused is not specified"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Move step_size to cpu if compiling in Adam and AdamW"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now move the cuda tensor to cpu if we're compiling.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Disable dynamo on functional optims if capturable=False"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now fallback to eager if capturable=False.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -118,6 +118,22 @@ def _dispatch_sqrt(
+def _disable_dynamo_if_unsupported(func):
+    def maybe_fallback(self, *args, **kwargs):
+        if is_compiling() and not kwargs.get("capturable", False):
+            import torch._dynamo
```
is this import still needed if you use @torch._disable_dynamo below?

mlazos replied:
Nope I was playing around with something else I think, let me remove.

janeyx99 commented on torch/optim/adamw.py:
```diff
@@ -670,6 +672,7 @@ def _multi_tensor_adamw(
 
             torch._foreach_div_(exp_avg_sq_sqrt, bias_correction2_sqrt)
             torch._foreach_add_(exp_avg_sq_sqrt, eps)
+
```
nit: could you undo this?

janeyx99 commented on torch/optim/adam.py:
```diff
@@ -677,6 +679,7 @@ def _multi_tensor_adam(
 
             torch._foreach_div_(exp_avg_sq_sqrt, bias_correction2_sqrt)
             torch._foreach_add_(exp_avg_sq_sqrt, eps)
+
```
nit: could you undo the space addition?

janeyx99 reviewed (approved):
Thanks, this looks good! Let some minor comments. >:( about the torchscript errors though, cc @albanD for how we should handle the failures?

mlazos commented:
> Thanks, this looks good! Let some minor comments. >:( about the torchscript errors though, cc @albanD for how we should handle the failures?

So to get around the torchscript thing I could do an if statement inside the functional optim where we call func, and disable if we're compiling and capturable=False. I think that would get around the varargs thing but it's kind of annoying. Lmk.

mlazos committed:
Update on "Disable dynamo on functional optims if capturable=False"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now fallback to eager if capturable=False.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Disable dynamo on functional optims if capturable=False"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now fallback to eager if capturable=False.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze




[ghstack-poisoned]

mlazos committed:
Update on "Disable dynamo on functional optims if capturable=False"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now fallback to eager if capturable=False.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze

Note on the actual fallback logic - there was an issue with torchscript originally not handling *args, **kwargs properly, after rectifying that by using `functools.wraps`, there was an additional bug with scoping which required the single tensor implementation to be in the global scope at the time of the fallback closure being created. I pass in the single tensor function to the `_disable_dynamo_if_unsupported` decorator to workaround this bug. 




[ghstack-poisoned]

mlazos commented:
> > Thanks, this looks good! Let some minor comments. >:( about the torchscript errors though, cc @albanD for how we should handle the failures?
> 
> So to get around the torchscript thing I could do an if statement inside the functional optim where we call func, and disable if we're compiling and capturable=False. I think that would get around the varargs thing but it's kind of annoying. Lmk.

I ended up being able to do the standard decorator but had to add the single tensor function to the surrounding environment where the decorator closure is created. This works now though without any TS BC-breaking changes.

mlazos commented:
> Why is moving the whole method to the end of the file needed?

The decorator takes in the single tensor function (which needs to be defined) to get around the TS bug I was talking about.

mlazos committed:
Update on "Disable dynamo on functional optims if capturable=False"

This resolves a bug in eager where if an old state dict is loaded (without the capturable flag) but the original dict had the capturable flag, then state_steps would be on cuda but we would take the non-capturable path. We now fallback to eager if capturable=False.

Current design doc and discussion: https://docs.google.com/document/d/1DmmbiaSp16CDZtGw1qzXKHFTY_0gqc0xpnBdviXq0vk/edit#heading=h.871u7bvwz7ze

Note on the actual fallback logic - there was an issue with torchscript originally not handling *args, **kwargs properly, after rectifying that by using `functools.wraps`, there was an additional bug with scoping which required the single tensor implementation to be in the global scope at the time of the fallback closure being created. I pass in the single tensor function to the `_disable_dynamo_if_unsupported` decorator to workaround this bug. 




[ghstack-poisoned]

albanD commented on torch/distributed/optim/__init__.py:
```diff
@@ -24,6 +24,9 @@
 from .utils import as_functional_optim
 
+from warnings import warn
+warn("TorchScript support for functional optimizers is"
```
@mlazos can you give some explanation of what that change is? I don't see it explained in the PR description?

This looks quite worrying because anyone that is importing torch.distributed.optim will now see this warning. Is your plan to remove this whole folder altogether? There are a lot of things here that have nothing to do with torchscipt or functional optimizers?

mlazos replied:
So every functional optimizer in that folder has the `torch.jit.script` decorator. Since this is the main way those are imported, rather than put the warning in each file I thought here was more optimal. I'll just remove the `torch.jit.script` decorator from each functional optim, not remove the whole folder. 

I can add this to the PR description, sorry I thought the intention was clear from our discussion.

albanD replied:
But there are many other things than these functional optimizers here?
In any case, I'll let @wconstab / @wanchaol  make the decision on the best way forward here since it is distributed code!
